### PR TITLE
✨ s3: allow best effort delete for objects

### LINKED ISSUE #4903: s3: allow best effort delete for objects
/kind feature

**Describe the solution you'd like**
Some users have to install clusters in accounts with policies that deny deleting S3 objects, e.g.
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "Deny object deletions",
            "Effect": "Deny",
            "Principal": "*",
            "Action": "s3:DeleteObject",
            "Resource": "arn:aws:s3:::<bucketname>/*"
        }
    ]
}
```
If you try to delete an object with that policy in place, you get:
```bash
 AccessDenied: Access Denied
	status code: 403, request id: RJKZ2D9H1EMMT0DZ, host id: qJNgwUBQKP1KMLkxa6WkaoN0z5bVHJzYzWWrd60HZ7SiYQcpB6h5BpDlSsZH/36t6tQoXUPbvdd3if7RPQxrvA==
```
Ideally such errors should not prevent installs from succeeding. To keep backward compatibility, I suggest a new config option for `S3Bucket`, something like `BestEffortDeleteObjects` which will cause CAPA to ignore `AccessDenied` errors when that option is present and true.

**Anything else you would like to add:**
A PR will follow.

**Environment:**

- Cluster-api-provider-aws version: main
- Kubernetes version: (use `kubectl version`): 1.29
- OS (e.g. from `/etc/os-release`): RHCOS

### PR DESCRIPTION
**Pull request by r4f4**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This change introduces a new config option for `S3Bucket` to allow best effort delete for S3 objects. This is useful to not fail cluster installs when the user has policies in place that prevent object deletion.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4903

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [X] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)
- [X] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Add a new field to S3Bucket, BestEffortDeleteObjects, to allow ignoring AccessDenied errors when deleting S3 objects.
```

### CODE REVIEW

**Comment by r4f4** [comment]
/retitle ✨ s3: allow best effort delete for objects

**Comment by r4f4** [comment]
CC @patrickdillon @mtulio

File: pkg/cloud/services/s3/s3.go [hunk 1]
```diff
@@ -221,11 +218,25 @@ func (s *Service) Delete(m *scope.MachineScope) error {
...
 		Key:    aws.String(key),
-	})
-	if err != nil {
+	}); err != nil {
+		if ptr.Deref(s.scope.Bucket().BestEffortDeleteObjects, false) {
+			if aerr, ok := err.(awserr.Error); ok {
+				switch aerr.Code() {
+				case "Forbidden":
+					fallthrough
+				case "AccessDenied":
```
Comment (Ankitasw) [inline pkg/cloud/services/s3/s3.go:234]: What's the difference between Forbidden and AccessDenied in this case? Why are we ignoring the AccessDenied case?

Comment (r4f4) [reply pkg/cloud/services/s3/s3.go:234]: No difference, we are treating them the same (that's why the `fallthrough`). But I'll change this as the linter suggests which should make it more clear.

**Comment by r4f4** [comment]
/hold
Until I squash the commits.

**Comment by r4f4** [comment]
@Ankitasw I've addressed your feedback points. Can you take another look?

**Comment by r4f4** [comment]
@richardcase any comments on this?

**Comment by richardcase** [comment]
From my side this looks good

/approve

Until the e2e pass:

/hold

**Review by damdo — COMMENTED** [review]
/lgtm

**Comment by r4f4** [comment]
@richardcase I think we're ready to remove the hold.

**Comment by damdo** [comment]
Considering the comment from @richardcase 
https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4904#issuecomment-2072325372

Given E2Es are green let's unhold this
/unhold

**Pull request merged**